libraries/AP_Mission/AP_Mission_ChangeDetector_Copter.cpp

```diff
@@ -0,0 +1,168 @@
+// using_next_command should be set to try if waypoint controller is already using the next navigation command
+AP_Mission_ChangeDetector_Copter::ChangeResponseType AP_Mission_ChangeDetector_Copter::check_for_mission_change(bool using_next_command)
+{
+    // take backup of command list
+    MissionCommandList cmd_list_bak = mis_change_detect;
+
+    uint8_t first_changed_cmd_idx = 0;
+    if (!AP_Mission_ChangeDetector::check_for_mission_change(first_changed_cmd_idx)) {
+        // the mission has not changed
+        return AP_Mission_ChangeDetector_Copter::ChangeResponseType::NONE;
+    }
+
+    // if the current command has change a reset is always required
+    // ToDo: check this handles mission erased
+    if (first_changed_cmd_idx == 0) {
+        debug("check_for_mission_change: 1st comment changed, Reset");
+        return AP_Mission_ChangeDetector_Copter::ChangeResponseType::RESET_REQUIRED;
+    }
+
+    // 1st exists and 2nd or 3rd command has been added, changed or deleted
+
+    const bool cmd0_is_wp = (mis_change_detect.cmd[0].id == MAV_CMD_NAV_WAYPOINT);
+    const bool cmd0_is_loiter = (mis_change_detect.cmd[0].id == MAV_CMD_NAV_LOITER_UNLIM ||
+                                 mis_change_detect.cmd[0].id == MAV_CMD_NAV_LOITER_TIME);
+    const bool cmd0_is_spline = (mis_change_detect.cmd[0].id == MAV_CMD_NAV_SPLINE_WAYPOINT);
+    const bool cmd0_is_other = (!cmd0_is_wp && !cmd0_is_loiter && !cmd0_is_spline);
+
+    // if 1st segment is not a wp, loiter or spline then return NONE
+    if (cmd0_is_other) {
+        debug("check_for_mission_change: 1st is neither wp nor spline, None");
+        return AP_Mission_ChangeDetector_Copter::ChangeResponseType::NONE;
+    }
+
+    const bool cmd0_has_pause = (mis_change_detect.cmd[0].p1 > 0);
+
+    // if 1st segment wp/spline has pause or is loiter then return NONE
+    if (((cmd0_is_wp || cmd0_is_spline) && cmd0_has_pause) || cmd0_is_loiter) {
+        debug("check_for_mission_change: 1st has pause, None");
+        return AP_Mission_ChangeDetector_Copter::ChangeResponseType::NONE;
+    }
+
+    const bool cmd1_exists = mis_change_detect.cmd_count >= 2;
+    const bool cmd1_has_pause = (mis_change_detect.cmd[1].p1 > 0);
+    const bool cmd1_is_wp = (mis_change_detect.cmd[1].id == MAV_CMD_NAV_WAYPOINT ||
+                             mis_change_detect.cmd[1].id == MAV_CMD_NAV_LOITER_UNLIM ||
+                             mis_change_detect.cmd[1].id == MAV_CMD_NAV_LOITER_TIME);
+    const bool cmd1_is_spline = (mis_change_detect.cmd[1].id == MAV_CMD_NAV_SPLINE_WAYPOINT);
+    const bool cmd1_is_other = (!cmd1_is_wp && !cmd1_is_spline);
+    const bool cmd1_existed = cmd_list_bak.cmd_count >= 2;
+    const bool cmd1_was_wp = (cmd_list_bak.cmd[1].id == MAV_CMD_NAV_WAYPOINT ||
+                             cmd_list_bak.cmd[1].id == MAV_CMD_NAV_LOITER_UNLIM ||
+                             cmd_list_bak.cmd[1].id == MAV_CMD_NAV_LOITER_TIME);
```

Comment:
this is the same RHS as cmd0_was_wp - a method on the wpnav library might work here, or perhaps a two-element array which is populated in a loop.
```
struct _wpinfo {
   bool exists;
  bool has_pause;
  bool is_wp;
} wpinfo[2];
```